pkg/controller/release/release_controller.go

```diff
@@ -283,30 +272,156 @@ func (c *Controller) syncOneReleaseHandler(key string) error {
+	}
+	if rel != nil && !equality.Semantic.DeepEqual(initialRel, rel) {
+		var updErr error
+		rel, updErr = c.clientset.ShipperV1alpha1().Releases(rel.Namespace).Update(rel)
+		if updErr != nil {
+			return shippererrors.NewKubeclientUpdateError(rel, updErr).WithShipperKind("Release")
+		}
+	}
+	if relStrategyErr != nil {
+		return relStrategyErr
+	}
+
+	for _, patch := range patches {
+		if err := c.applyPatch(namespace, patch); err != nil {
+			return err
 		}
 	}
 
+	klog.V(4).Infof("Done processing Release %q", key)
+
+	return nil
+}
+
+func (c *Controller) applicationReleases(rel *shipper.Release) ([]*shipper.Release, error) {
+	appName, err := releaseutil.ApplicationNameForRelease(rel)
+	if err != nil {
+		return nil, err
+	}
+	releases, err := c.releaseLister.Releases(rel.Namespace).ReleasesForApplication(appName)
+	if err != nil {
+		return nil, err
+	}
+	return releases, nil
+}
+
+func (c *Controller) ensureReleaseState(rel *shipper.Release) (*shipper.Release, []ExecutorResult, []ReleaseStrategyStateTransition, error) {
+	diff := diffutil.NewMultiDiff()
+	defer func() {
+		if !diff.IsEmpty() {
+			c.recorder.Eventf(rel, corev1.EventTypeNormal, "ReleaseConditionChanged", diff.String())
+		}
+	}()
+	releases, err := c.applicationReleases(rel)
 	if err != nil {
-		return err
+		return rel, nil, nil, err
+	}
+	prev, succ, err := releaseutil.GetSiblingReleases(rel, releases)
+	if err != nil {
+		return rel, nil, nil, err
 	}
 
-	appKey, err := c.getAssociatedApplicationKey(rel)
+	relInfoCurr, err := c.buildReleaseInfo(rel)
 	if err != nil {
-		return err
+		return rel, nil, nil, err
+	}
+	var relInfoPrev, relInfoSucc *releaseInfo
+	if prev != nil {
+		relInfoPrev, err = c.buildReleaseInfo(prev)
+		if err != nil {
+			return rel, nil, nil, err
+		}
+	}
+	if succ != nil {
+		relInfoSucc, err = c.buildReleaseInfo(succ)
+		if err != nil {
+			return rel, nil, nil, err
+		}
 	}
 
-	// If everything went fine, scheduling an application key in the
-	// application workqueue.
-	klog.V(4).Infof("Scheduling Application key %q", appKey)
-	c.applicationWorkqueue.Add(appKey)
+	//TODO: this flag is deprecated, we keep it here for the sake of gradual
+	//changes. The flag propagates as deep as strategy conditions
+	//consolidation functions
+	// see pkg/util/conditions/strategy.go for more details.
+	hasIncumbent := len(releases) > 1
+	executor := NewStrategyExecutor(relInfoCurr, relInfoPrev, relInfoSucc, c.recorder, hasIncumbent)
 
-	klog.V(4).Infof("Done processing Release %q", key)
+	patches, trans, err := executor.Execute()
+	if err != nil {
+		releaseStrategyExecutedCond := releaseutil.NewReleaseCondition(
+			shipper.ReleaseConditionTypeStrategyExecuted,
+			corev1.ConditionFalse,
+			conditions.StrategyExecutionFailed,
+			fmt.Sprintf("failed to execute strategy: %q", err),
+		)
+		diff.Append(releaseutil.SetReleaseCondition(&rel.Status, *releaseStrategyExecutedCond))
+		return rel, nil, nil, err
+	}
+
+	if !releaseutil.ReleaseStrategyExecuted(rel) {
```

Comment:
I'd rather set the condition unconditionally. Having a conditional around has proven to attract problems in the past (see #125 for an example), and also goes against our philosophy of not reading conditions that we own (pretty much for the same reasons).